File: java/PermutationSequence.java
public class Solution {
    public String getPermutation(int n, int k) {
        StringBuilder result = new StringBuilder();
        List<Integer> list = new ArrayList<Integer>();
        int category = 1;
        for (int i = 1; i <= n; i++) {
            list.add(i);
            if (i != n) {
                category = category * i;
            }
        }

        k = k - 1;
        while (result.length() < n) {
            int index = k / category;
            result.append(list.get(index));
            list.remove(index);

            k = k % category;
            if (n != result.length()) {
                category = category / (n - result.length());
            }
        }

        return result.toString();
    }
}

File: archive/python/math/PermutationSequence.py
class Solution:
    # @return a string
    def getPermutation(self, n, k):
        if n == 0:
            return "0"
            
        num_list = []
        for i in xrange(n):
            num_list.append(str(i+1))
            
        total = 1
        result = ""
        for i in xrange(n):
            total *= (i+1)
            
        for i in xrange(n, 0, -1):
            count = total / i
            index = (k - 1) / count
            result += num_list[index]
            total = count
            k = k - (count * index)
            num_list = num_list[0:index] + num_list[index+1:]
            
        return result

if __name__ == "__main__":
    solution = Solution()

    print solution.getPermutation(4, 6)
